tests/tools/fswalker/fswalker_test.go

```diff
@@ -0,0 +1,533 @@
+		}
+
+		// If an error was thrown, expect a report to be written to the provided writer
+		if (reportOut.Len() > 0) != tt.wantErr {
+			t.Errorf("report length unexpected len = %v, expReport %v", reportOut.Len(), tt.wantErr)
+		}
+	}
+}
+
+func TestWalkChecker_filterReportDiffs(t *testing.T) {
+	type fields struct {
+		GlobalFilterMatchers []string
+	}
+
+	for _, tt := range []struct {
+		name        string
+		fields      fields
+		inputReport *fswalker.Report
+		expModCount int
+	}{
+		{
+			name: "No filters",
+			fields: fields{
+				GlobalFilterMatchers: nil,
+			},
+			inputReport: &fswalker.Report{
+				Modified: []fswalker.ActionData{
+					{
+						Diff: "some difference",
+					},
+				},
+			},
+			expModCount: 1,
+		},
+		{
+			name: "filter the only diff as prefix",
+			fields: fields{
+				GlobalFilterMatchers: []string{
+					"some",
+				},
+			},
+			inputReport: &fswalker.Report{
+				Modified: []fswalker.ActionData{
+					{
+						Diff: "some difference",
+					},
+				},
+			},
+			expModCount: 0,
+		},
+		{
+			name: "filter the only diff as substring",
+			fields: fields{
+				GlobalFilterMatchers: []string{
+					"iff",
+				},
+			},
+			inputReport: &fswalker.Report{
+				Modified: []fswalker.ActionData{
+					{
+						Diff: "some difference",
+					},
+				},
+			},
+			expModCount: 0,
+		},
+		{
+			name: "filter some but not all diffs",
+			fields: fields{
+				GlobalFilterMatchers: []string{
+					"definitely",
+				},
+			},
+			inputReport: &fswalker.Report{
+				Modified: []fswalker.ActionData{
+					{
+						Diff: "this will not be filtered",
+					},
+					{
+						Diff: "this will definitely be filtered",
+					},
+				},
+			},
+			expModCount: 1,
+		},
+		{
+			name: "filter multiple diffs",
+			fields: fields{
+				GlobalFilterMatchers: []string{
+					"definitely",
+				},
+			},
+			inputReport: &fswalker.Report{
+				Modified: []fswalker.ActionData{
+					{
+						Diff: "this will not be filtered",
+					},
+					{
+						Diff: "this will definitely be filtered",
+					},
+					{
+						Diff: "this will also definitely be filtered",
+					},
+				},
+			},
+			expModCount: 1,
+		},
+	} {
+		t.Log(tt.name)
+
+		chk := &WalkCompare{
+			GlobalFilterMatchers: tt.fields.GlobalFilterMatchers,
+		}
+		chk.filterReportDiffs(tt.inputReport)
+
+		if want, got := tt.expModCount, len(tt.inputReport.Modified); want != got {
+			t.Errorf("Expected %v modifications after filter but got %v (%v)", want, got, tt.inputReport.Modified)
+		}
+	}
+}
+
+func Test_isRootDirectoryRename(t *testing.T) {
+	type args struct {
+		diffItem string
+		mod      fswalker.ActionData
+	}
+
+	for _, tt := range []struct {
+		name string
+		args args
+		want bool
+	}{
+		{
+			name: "Check a root name change",
+			args: args{
+				diffItem: "name: \"fio-data-902268402\" => \"restore-snap-43720e98eaa9b40ec0be735e347bb853964221402\"",
+				mod: fswalker.ActionData{
+					Before: &fspb.File{
+						Path: "this_is_restore_directory_root",
```

Comment:
Does path need to be empty? I am confused about the semantics of `isRootDirectoryRename` and when it happen.

Reply:
oh so, `rerootWalkDataPaths` is called after each walk. This modifies the paths inside the walk to all be relative to the walk root (i.e. where the walk started). This is so `/some/path/to/source/file.txt` will have the same path as `/path/leading/to/target/file.txt`

Reply:
because in both cases we just care that `file.txt` exists in the directory of the snapshot/restore itself, not what its full path is

Reply:
so after `rerootWalkDataPaths` the root directory path becomes ".", and the file's path becomes `./file.txt`, which is comparable in both before and after. `isRootDirectoryRename` prevents the reporter from flagging an error on that root directory entry, which before was `source`, and after is `target`. It sees it's the root directory and expects the name could have changed, like with the kopia commands I posted in the other comment

Reply:
Let me double check on this... make sure it's doing what I'm expecting

Reply:
ok so seems I'm missing a test case and there's a bug (thanks for making me revisit this). I should be expecting the path of the root is `"."`, but currently I'm checking that the _parent directory_ of the root is ".", which applies to both the root directory itself (whose path and parent are both ".") and a subdirectory (whose parent will be ".")

Reply:
Alrighty added some more tests/test cases and fixed the bug. Thanks again @julio-lopez